### PR TITLE
[17.0][OU-ADD] account: migration to 17.0

### PR DESCRIPTION
**Pull request by daiduongnguyen-odoo**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
/ocabot migration account

<dependency>Depends on : 

- product : #4598
- analytic : #4597
- digest : #4642</dependency>

**Comment by daiduongnguyen-odoo** [comment]
@flachica Hello, i apply your fix here, please notify me if you have something in your mind :v

**Comment by flachica** [comment]
> @flachica Hello, i apply your fix here, please notify me if you have something in your mind :v

I will close my PR and internally use this branch. Thanks for you amazing work.

**Review by flachica — APPROVED** [review]

**Comment by rvalyi** [comment]
Hello, I'm creating a new analysis tool called odoo-module-diff, it's explained here: https://github.com/akretion/odoo-module-diff

To review or complete the PR, you might be interested by these key commits of the account module migration to 17.0:
https://github.com/akretion/odoo-module-diff-analysis/tree/main/17.0/account

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (ecino) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:17]: When trying the script, I get the following error:

```
psycopg2.errors.DatatypeMismatch: column "invoice_label" is of type jsonb but expression is of type character varying
LINE 4: ...             UPDATE account_tax SET invoice_label=description
```

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (daiduongnguyen-odoo) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:605]: @ecino Ok maybe do same here

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:213]: as we don't do anything with this ourselves, we should leave it to the ORM to create this column. It will be available to the depending modules anyways

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:227]: see above

File: openupgrade_scripts/scripts/account/17.0.1.2/upgrade_analysis_work.txt
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/upgrade_analysis_work.txt:129]: this and the above should be nothing to do

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:287]:
```suggestion
                WHEN end_month IS NULL OR NOT end_month THEN 'days_after'
```

but given the ELSE branch below does the same, we can just remove this

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:286]:
```suggestion
                WHEN end_month = true AND COALESCE(months, 0) = 0 AND days_after > 0 THEN 'days_end_of_month_on_the'
                WHEN end_month = true AND COALESCE(months, 0) = 0 THEN 'days_after_end_of_month'
                WHEN end_month = true AND months > 1 THEN 'days_end_of_month_on_the'
```

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:297]: we can't map all possible configurations of v16 here, so cover the two cases that match exactly, do a heuristic that might be what the user wants for the rest

```suggestion
                WHEN end_month = true AND months <= 1
                THEN COALESCE(days, 0) + COALESCE(days_after, 0)
                ELSE COALESCE(months, 0)*30 + COALESCE(days, 0) + COALESCE(days_after, 0)
```

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:249]: this function should go to post-migration to avoid having to create fields manually

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:345]: do this conditionally if we set `delay_type=days_end_of_month_on_the` above, pre-create column  `days_next_month`(https://github.com/OCA/OCB/blob/17.0/addons/account_payment_term/models/account_payment_term.py#L16) and set it to `days_after`, subtract from `days` as we add it above in those lines too

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:151]: move to post-migration

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:130]: move to post-migration, use [openupgrade.map_values](https://oca.github.io/openupgradelib/API.html#openupgradelib.openupgrade.map_values)

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:191]: why this restriction?

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:604]: in the spirit of preserving data, I'd rather copy `description` with type `None`, then run `convert_column_translatable`on the result

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:432]: this code will be much simpler in post-migrate. I also suggest to add a column in tax groups pointing to the tax group they originate from, this will make it simpler for localizations to find them and do something with them if necessary

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:478]: better clone this one and write on the clone, then it's less hassle for localization to process them

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:534]: third case is both company_id and res_id are null

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:515]: I think it's useful to add `openupgrade.convert_from_company_dependent` which is the inverse of [`convert_to_company_dependent`](https://oca.github.io/openupgradelib/API.html#openupgradelib.openupgrade.convert_to_company_dependent)

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:402]: remove

File: openupgrade_scripts/scripts/account/17.0.1.2/upgrade_analysis_work.txt
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/upgrade_analysis_work.txt:364]: should be nothing to do

File: openupgrade_scripts/scripts/account/17.0.1.2/end-migration.py
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/end-migration.py:1]: remove whole file

File: openupgrade_scripts/scripts/account/17.0.1.2/upgrade_analysis_work.txt
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/upgrade_analysis_work.txt:374]: as written above, I think we should keep them

File: openupgrade_scripts/scripts/account/17.0.1.2/upgrade_analysis_work.txt
Comment (hbrunn) [inline openupgrade_scripts/scripts/account/17.0.1.2/upgrade_analysis_work.txt:423]: this one is actually added in account now, so we need to uniquify names.

And the whole _work file needs updating from the current analysis file, it contains confusing bogus entries

**Review by hbrunn — CHANGES_REQUESTED** [review]

**Comment by hbrunn** [comment]
I'll create a followup-PR next Monday of nobody else has picked this up until then

**Comment by daiduongnguyen-odoo** [comment]
> I'll create a followup-PR next Monday of nobody else has picked this up until then

Thank you sir

**Comment by hbrunn** [comment]
I didn't manage to finish this today, so I just pushed an [intermediate version](https://github.com/hbrunn/OpenUpgrade/tree/17.0-account) on my fork. ~~Still working on a correct payment term migration,~~ Payment terms and tax groups should be fine already.

**Comment by pedrobaeza** [comment]
@hbrunn is this finished on your side (or on hold)? It's for resuming from here.

**Comment by hbrunn** [comment]
my comments from https://github.com/OCA/OpenUpgrade/pull/4457#discussion_r1819149393 down need addressing still, so if you've the time, please resume from there

**Comment by hbrunn** [comment]
PS: I'll continue with this on Monday if I don't hear anything to the contrary

**Comment by pedrobaeza** [comment]
I'm today resuming this, so probably it will be done. I'll tell you.

File: openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py
Comment (pedrobaeza) [reply openupgrade_scripts/scripts/account/17.0.1.2/pre-migration.py:213]: @hbrunn It's OK to pre-create this column for avoiding to trigger the computation (although empty in this module), so we light the process. I'm reverting the change you have done.

**Comment by hbrunn** [comment]
I don't find any new commits on https://github.com/Tecnativa/OpenUpgrade/commits/17.0-account - what are you working on so that I can do something else today?

**Comment by pedrobaeza** [comment]
I'm still working on this.

**Comment by hbrunn** [comment]
today?

**Comment by pedrobaeza** [comment]
Yes, I will push my result at the end of the morning.

**Comment by hbrunn** [comment]
great, then I'll pick it up from there and do some more reviews meanwhile

**Comment by hbrunn** [comment]
?

**Comment by pedrobaeza** [comment]
I have opened https://github.com/OCA/OpenUpgrade/pull/4662, but it's still not finished. I'm just stuck at the new `invoice_label` to fill proper data. I have reviewed/reworked till https://github.com/OCA/OpenUpgrade/pull/4662/files#diff-22f22a0c72f96258826774abe5ffb855f21883cf30e51b7e81fbcaaca048214aR275

**Pull request closed without merging**
